File: src/unicon/StateEncoder.cpp
#include "StateEncoder.h"

#include <SDL3/SDL_log.h>
#include <glm/gtc/quaternion.hpp>
#include <algorithm>
#include <cmath>

void StateEncoder::configure(size_t numJoints, size_t targetFrameCount) {
    numJoints_ = numJoints;
    tau_ = targetFrameCount;

    SDL_Log("StateEncoder configured: %zu joints, tau=%zu, observation dim=%zu",
            numJoints_, tau_, getObservationDim());
}

size_t StateEncoder::getFrameEncodingDim() const {
    // Root height (1) + root quat (4) + joint positions (3J) + joint quats (4J)
    // + root lin vel (3) + root ang vel (3) + joint ang vels (3J)
    return 11 + 10 * numJoints_;
}

size_t StateEncoder::getObservationDim() const {
    // (1 + tau) frame encodings + tau root offsets
    return (1 + tau_) * getFrameEncodingDim() + tau_ * ROOT_OFFSET_DIM;
}

void StateEncoder::encode(const ArticulatedBody& body,
                           const PhysicsWorld& physics,
                           const std::vector<TargetFrame>& targetFrames,
                           std::vector<float>& observation) const {
    const size_t dim = getObservationDim();
    observation.resize(dim, 0.0f);

    if (body.getPartCount() == 0 || numJoints_ == 0) return;

    // Extract current physics state
    std::vector<ArticulatedBody::PartState> states;
    body.getState(states, physics);

    const size_t partCount = std::min(states.size(), numJoints_);

    // Current root state
    const glm::vec3& rootPos = states[0].position;
    const glm::quat& rootRot = states[0].rotation;
    const glm::vec3& rootLinVel = states[0].linearVelocity;
    const glm::vec3& rootAngVel = states[0].angularVelocity;

    // Gather joint positions, rotations, angular velocities from physics
    std::vector<glm::vec3> jointPositions(numJoints_);
    std::vector<glm::quat> jointRotations(numJoints_);
    std::vector<glm::vec3> jointAngVels(numJoints_);

    for (size_t i = 0; i < partCount; ++i) {
        jointPositions[i] = states[i].position;
        jointRotations[i] = states[i].rotation;
        jointAngVels[i] = states[i].angularVelocity;
    }

    float* out = observation.data();

    // Encode current state o(X_t)
    size_t written = encodeFrame(rootPos, rootRot, rootLinVel, rootAngVel,
                                 jointPositions, jointRotations, jointAngVels, out);
    out += written;

    // Encode each target frame o(X~_{t+k}) for k = 1..tau
    const size_t numTargets = std::min(targetFrames.size(), tau_);
    for (size_t k = 0; k < numTargets; ++k) {
        const auto& target = targetFrames[k];

        // Pad target joint arrays if smaller than numJoints_
        std::vector<glm::vec3> tJointPos(numJoints_, glm::vec3(0.0f));
        std::vector<glm::quat> tJointRot(numJoints_, glm::quat(1.0f, 0.0f, 0.0f, 0.0f));
        std::vector<glm::vec3> tJointAngVel(numJoints_, glm::vec3(0.0f));

        size_t count = std::min(target.jointPositions.size(), numJoints_);
        for (size_t i = 0; i < count; ++i) tJointPos[i] = target.jointPositions[i];

        count = std::min(target.jointRotations.size(), numJoints_);
        for (size_t i = 0; i < count; ++i) tJointRot[i] = target.jointRotations[i];

        count = std::min(target.jointAngularVelocities.size(), numJoints_);
        for (size_t i = 0; i < count; ++i) tJointAngVel[i] = target.jointAngularVelocities[i];

        written = encodeFrame(target.rootPosition, target.rootRotation,
                              target.rootLinearVelocity, target.rootAngularVelocity,
                              tJointPos, tJointRot, tJointAngVel, out);
        out += written;
    }

    // Zero-pad any missing target frames
    for (size_t k = numTargets; k < tau_; ++k) {
        std::fill_n(out, getFrameEncodingDim(), 0.0f);
        out += getFrameEncodingDim();
    }

    // Encode root offsets y(X_t, X~_{t+k}) for k = 1..tau
    for (size_t k = 0; k < numTargets; ++k) {
        const auto& target = targetFrames[k];
        written = encodeRootOffset(rootPos, rootRot,
                                   target.rootPosition, target.rootRotation, out);
        out += written;
    }

    // Zero-pad any missing root offsets
    for (size_t k = numTargets; k < tau_; ++k) {
        std::fill_n(out, ROOT_OFFSET_DIM, 0.0f);
        out += ROOT_OFFSET_DIM;
    }
}

size_t StateEncoder::encodeFrame(const glm::vec3& rootPos, const glm::quat& rootRot,
                                  const glm::vec3& rootLinVel, const glm::vec3& rootAngVel,
                                  const std::vector<glm::vec3>& jointPositions,
                                  const std::vector<glm::quat>& jointRotations,
                                  const std::vector<glm::vec3>& jointAngVels,
                                  float* out) const {
    float* start = out;

    // Compute inverse root rotation for transforming to root-local frame.
    // Root heading = rotation around Y axis only (remove pitch/roll).
    float yaw = atan2f(2.0f * (rootRot.w * rootRot.y + rootRot.x * rootRot.z),
                       1.0f - 2.0f * (rootRot.y * rootRot.y + rootRot.z * rootRot.z));
    glm::quat headingRot = glm::angleAxis(yaw, glm::vec3(0.0f, 1.0f, 0.0f));
    glm::quat headingInv = glm::inverse(headingRot);

    // 1. Root height (Y-up)
    *out++ = rootPos.y;

    // 2. Root rotation quaternion in heading-local frame
    // This captures pitch/roll relative to heading direction
    glm::quat localRot = headingInv * rootRot;
    *out++ = localRot.w;
    *out++ = localRot.x;
    *out++ = localRot.y;
    *out++ = localRot.z;

    // 3. Joint positions relative to root, rotated into heading frame (3J)
    for (size_t i = 0; i < numJoints_; ++i) {
        glm::vec3 relPos = headingInv * (jointPositions[i] - rootPos);
        *out++ = relPos.x;
        *out++ = relPos.y;
        *out++ = relPos.z;
    }

    // 4. Joint rotation quaternions in heading-local frame (4J)
    for (size_t i = 0; i < numJoints_; ++i) {
        glm::quat localJointRot = headingInv * jointRotations[i];
        *out++ = localJointRot.w;
        *out++ = localJointRot.x;
        *out++ = localJointRot.y;
        *out++ = localJointRot.z;
    }

    // 5. Root linear velocity in heading-local frame (3)
    glm::vec3 localLinVel = headingInv * rootLinVel;
    *out++ = localLinVel.x;
    *out++ = localLinVel.y;
    *out++ = localLinVel.z;

    // 6. Root angular velocity in heading-local frame (3)
    glm::vec3 localAngVel = headingInv * rootAngVel;
    *out++ = localAngVel.x;
    *out++ = localAngVel.y;
    *out++ = localAngVel.z;

    // 7. Joint angular velocities in heading-local frame (3J)
    for (size_t i = 0; i < numJoints_; ++i) {
        glm::vec3 localJAngVel = headingInv * jointAngVels[i];
        *out++ = localJAngVel.x;
        *out++ = localJAngVel.y;
        *out++ = localJAngVel.z;
    }

    return static_cast<size_t>(out - start);
}

size_t StateEncoder::encodeRootOffset(const glm::vec3& actualRootPos, const glm::quat& actualRootRot,
                                       const glm::vec3& targetRootPos, const glm::quat& targetRootRot,
                                       float* out) const {
    // Transform position offset into actual root's heading-local frame
    float yaw = atan2f(2.0f * (actualRootRot.w * actualRootRot.y + actualRootRot.x * actualRootRot.z),
                       1.0f - 2.0f * (actualRootRot.y * actualRootRot.y + actualRootRot.z * actualRootRot.z));
    glm::quat headingRot = glm::angleAxis(yaw, glm::vec3(0.0f, 1.0f, 0.0f));
    glm::quat headingInv = glm::inverse(headingRot);

    glm::vec3 worldOffset = targetRootPos - actualRootPos;
    glm::vec3 localOffset = headingInv * worldOffset;

    // 1. Horizontal position offset in root local frame (2: X, Z)
    *out++ = localOffset.x;
    *out++ = localOffset.z;

    // 2. Height offset (1)
    *out++ = localOffset.y;

    // 3. Rotation offset quaternion (4)
    // q_offset = inv(actual_heading) * target_rotation
    glm::quat rotOffset = headingInv * targetRootRot;
    *out++ = rotOffset.w;
    *out++ = rotOffset.x;
    *out++ = rotOffset.y;
    *out++ = rotOffset.z;

    return ROOT_OFFSET_DIM;
}

File: tools/unicon_training/state_encoder.py
"""Python StateEncoder matching the C++ implementation in src/unicon/StateEncoder.cpp.

Encodes character state as an observation vector for the MLP policy.
Uses the same heading-local coordinate frame and component ordering
so that trained weights transfer directly to the C++ runtime.
"""

import numpy as np
from typing import List, Optional


def quat_yaw(q: np.ndarray) -> float:
    """Extract yaw angle from quaternion (w,x,y,z format, Y-up)."""
    w, x, y, z = q
    return np.arctan2(2.0 * (w * y + x * z), 1.0 - 2.0 * (y * y + z * z))


def quat_from_axis_angle(axis: np.ndarray, angle: float) -> np.ndarray:
    """Create quaternion (w,x,y,z) from axis-angle."""
    half = angle * 0.5
    s = np.sin(half)
    return np.array([np.cos(half), axis[0] * s, axis[1] * s, axis[2] * s])


def quat_inverse(q: np.ndarray) -> np.ndarray:
    """Inverse of unit quaternion (w,x,y,z)."""
    return np.array([q[0], -q[1], -q[2], -q[3]])


def quat_mul(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    """Quaternion multiplication (w,x,y,z format)."""
    aw, ax, ay, az = a
    bw, bx, by, bz = b
    return np.array([
        aw * bw - ax * bx - ay * by - az * bz,
        aw * bx + ax * bw + ay * bz - az * by,
        aw * by - ax * bz + ay * bw + az * bx,
        aw * bz + ax * by - ay * bx + az * bw,
    ])


def quat_rotate_vec(q: np.ndarray, v: np.ndarray) -> np.ndarray:
    """Rotate vector v by quaternion q (w,x,y,z)."""
    qv = np.array([0.0, v[0], v[1], v[2]])
    result = quat_mul(quat_mul(q, qv), quat_inverse(q))
    return result[1:4]


class StateEncoder:
    """Builds observation vectors matching the C++ StateEncoder exactly.

    Observation layout:
        [o(X_t), o(X~_{t+1}), ..., o(X~_{t+tau}), y(X_t, X~_{t+1}), ..., y(X_t, X~_{t+tau})]

    Where o(X) encodes in heading-local frame:
        - Root height (1)
        - Root rotation quaternion (4)
        - Joint positions relative to root (3*J)
        - Joint rotation quaternions (4*J)
        - Root linear velocity (3)
        - Root angular velocity (3)
        - Joint angular velocities (3*J)
    Total per frame: 11 + 10*J

    Where y(X, X~) encodes root offset:
        - Horizontal offset x,z (2)
        - Height offset y (1)
        - Rotation offset quaternion (4)
    Total per offset: 7
    """

    ROOT_OFFSET_DIM = 7

    def __init__(self, num_joints: int, tau: int = 1):
        self.num_joints = num_joints
        self.tau = tau

    @property
    def frame_encoding_dim(self) -> int:
        return 11 + 10 * self.num_joints

    @property
    def observation_dim(self) -> int:
        return (1 + self.tau) * self.frame_encoding_dim + self.tau * self.ROOT_OFFSET_DIM

    def encode(
        self,
        root_pos: np.ndarray,
        root_rot: np.ndarray,
        root_lin_vel: np.ndarray,
        root_ang_vel: np.ndarray,
        joint_positions: np.ndarray,
        joint_rotations: np.ndarray,
        joint_ang_vels: np.ndarray,
        target_frames: List[dict],
    ) -> np.ndarray:
        """Build the full observation vector.

        Args:
            root_pos: (3,) root world position
            root_rot: (4,) root rotation quaternion (w,x,y,z)
            root_lin_vel: (3,) root linear velocity
            root_ang_vel: (3,) root angular velocity
            joint_positions: (J, 3) world-space joint positions
            joint_rotations: (J, 4) world-space joint rotations (w,x,y,z)
            joint_ang_vels: (J, 3) world-space joint angular velocities
            target_frames: list of tau dicts, each with keys:
                root_pos, root_rot, root_lin_vel, root_ang_vel,
                joint_positions, joint_rotations, joint_ang_vels

        Returns:
            observation: (observation_dim,) float32 array
        """
        obs = np.zeros(self.observation_dim, dtype=np.float32)
        offset = 0

        # Encode current state o(X_t)
        n = self._encode_frame(
            root_pos, root_rot, root_lin_vel, root_ang_vel,
            joint_positions, joint_rotations, joint_ang_vels,
            obs, offset,
        )
        offset += n

        # Encode target frames o(X~_{t+k})
        for k in range(self.tau):
            if k < len(target_frames):
                tf = target_frames[k]
                n = self._encode_frame(
                    tf["root_pos"], tf["root_rot"],
                    tf["root_lin_vel"], tf["root_ang_vel"],
                    tf["joint_positions"], tf["joint_rotations"],
                    tf["joint_ang_vels"],
                    obs, offset,
                )
            else:
                n = self.frame_encoding_dim
            offset += n

        # Encode root offsets y(X_t, X~_{t+k})
        for k in range(self.tau):
            if k < len(target_frames):
                tf = target_frames[k]
                n = self._encode_root_offset(
                    root_pos, root_rot,
                    tf["root_pos"], tf["root_rot"],
                    obs, offset,
                )
            else:
                n = self.ROOT_OFFSET_DIM
            offset += n

        return obs

    def _encode_frame(
        self,
        root_pos: np.ndarray,
        root_rot: np.ndarray,
        root_lin_vel: np.ndarray,
        root_ang_vel: np.ndarray,
        joint_positions: np.ndarray,
        joint_rotations: np.ndarray,
        joint_ang_vels: np.ndarray,
        out: np.ndarray,
        offset: int,
    ) -> int:
        """Encode a single frame into heading-local coordinates.
        Matches C++ StateEncoder::encodeFrame exactly."""
        idx = offset

        # Heading rotation (yaw only around Y axis)
        yaw = quat_yaw(root_rot)
        heading_rot = quat_from_axis_angle(np.array([0.0, 1.0, 0.0]), yaw)
        heading_inv = quat_inverse(heading_rot)

        # 1. Root height
        out[idx] = root_pos[1]  # Y-up
        idx += 1

        # 2. Root rotation in heading-local frame
        local_rot = quat_mul(heading_inv, root_rot)
        out[idx:idx + 4] = local_rot
        idx += 4

        # 3. Joint positions relative to root in heading frame (3*J)
        for i in range(self.num_joints):
            rel_pos = quat_rotate_vec(heading_inv, joint_positions[i] - root_pos)
            out[idx:idx + 3] = rel_pos
            idx += 3

        # 4. Joint rotation quaternions in heading frame (4*J)
        for i in range(self.num_joints):
            local_jrot = quat_mul(heading_inv, joint_rotations[i])
            out[idx:idx + 4] = local_jrot
            idx += 4

        # 5. Root linear velocity in heading frame
        local_lin_vel = quat_rotate_vec(heading_inv, root_lin_vel)
        out[idx:idx + 3] = local_lin_vel
        idx += 3

        # 6. Root angular velocity in heading frame
        local_ang_vel = quat_rotate_vec(heading_inv, root_ang_vel)
        out[idx:idx + 3] = local_ang_vel
        idx += 3

        # 7. Joint angular velocities in heading frame (3*J)
        for i in range(self.num_joints):
            local_jav = quat_rotate_vec(heading_inv, joint_ang_vels[i])
            out[idx:idx + 3] = local_jav
            idx += 3

        return idx - offset

    def _encode_root_offset(
        self,
        actual_pos: np.ndarray,
        actual_rot: np.ndarray,
        target_pos: np.ndarray,
        target_rot: np.ndarray,
        out: np.ndarray,
        offset: int,
    ) -> int:
        """Encode root offset y(X, X~).
        Matches C++ StateEncoder::encodeRootOffset exactly."""
        idx = offset

        yaw = quat_yaw(actual_rot)
        heading_rot = quat_from_axis_angle(np.array([0.0, 1.0, 0.0]), yaw)
        heading_inv = quat_inverse(heading_rot)

        world_offset = target_pos - actual_pos
        local_offset = quat_rotate_vec(heading_inv, world_offset)

        # 1. Horizontal offset (x, z)
        out[idx] = local_offset[0]
        out[idx + 1] = local_offset[2]
        idx += 2

        # 2. Height offset (y)
        out[idx] = local_offset[1]
        idx += 1

        # 3. Rotation offset quaternion
        rot_offset = quat_mul(heading_inv, target_rot)
        out[idx:idx + 4] = rot_offset
        idx += 4

        return self.ROOT_OFFSET_DIM
